### PR TITLE
new qmk subcommand for creating `compile_commands.json` for better clangd/IDE interop

### PR DESCRIPTION
**Pull request by xton**
## Description

(I'm not able to re-open https://github.com/qmk/qmk_firmware/pull/8916, so I'm filing a new PR)

I use clangd as a languageserver along with coc.vim[0]. It requires a compilation database[1] to help it figure out how to find all the required files for a given compilation pass. This PR adds a new subcommand (`compiledb`) which creates that for you. Now your dev environment can figure out what `#include QMK_KEYBOARD_H` means.

[0] https://github.com/neoclide/coc.nvim/wiki/Language-servers#ccobjective-c
[1] https://clang.llvm.org/docs/JSONCompilationDatabase.html

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/qmk/qmk_firmware/issues/4855

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by xton** [comment]
(See https://github.com/qmk/qmk_firmware/pull/8916 for prior discussion, feedback, and fixes)

**Comment by drashna** [comment]
two questions. I think I've asked this before, actually ...

1. This requires version 12 of clang, correct?  If so, that may be a problem. I know that MacOS is limited to 11.0.3 on catalina. 
2. This applies to only one keyboard, at a time, correct? I'm pretty sure it's "yes", based on code and comments, but want to be 100% sure.

**Comment by xton** [comment]
1. No, I'm using this with version 10 on mojave and it works just fine.
2. Yes. Just one at a time. It's quite quick to rebuild though.

**Comment by xton** [comment]
(also to be clear, I'm not using clang to build qmk. this is strictly used by my dev environment to help intellisense)

**Comment by drashna** [comment]
Okay, sounds good then.  I just wanted to make sure.

**Review by drashna — APPROVED** [review]

**Comment by MForster** [comment]
Just some quick feedback: This seems to work well for me, and I'd like to see this merged into master.

File: .gitignore [hunk 1]
```diff
@@ -76,3 +76,7 @@ __pycache__
 
 # Allow to exist but don't include it in the repo
 user_song_list.h
+
+# clangd
+compile_commands.json
+.clangd/
```
Comment (MForster) [inline .gitignore:82]: Newer versions of clangd store the index in `.clangd/`:

```suggestion
# clangd
compile_commands.json
.clangd/
.cache/
```

**Comment by xton** [comment]
committed suggestion.
ptal @MForster

**Comment by MForster** [comment]
> committed suggestion.
> ptal @MForster

lgtm

File: docs/cli_commands.md [hunk 2]
```diff
@@ -237,6 +237,27 @@ qmk cformat
...
+## `qmk compiledb`
+
+Creates a `compile_commands.json` file.
+
+Does your IDE/editor use a language server but doesn't _quite_ find all the necessary include files? Do you hate red squigglies? Do you wish your editor could figure out `#include QMK_KEYBOARD_H`? You might need a [compilation database](https://clang.llvm.org/docs/JSONCompilationDatabase.html)! The qmk tool can build this for you.
+
+This command needs to know which keyboard and keymap to build. It uses the same configuration options as the `qmk compile` command: arguments, current directory, and config files.
+
+**Example:**
+
```
Comment (skullydazed) [inline docs/cli_commands.md:249]:
```suggestion

**Usage**:
```
qmk compiledb [-kb KEYBOARD] [-km KEYMAP]
```

**Comment by xton** [comment]
Usage line added. Anything else?

ptal @skullydazed

File: lib/python/qmk/cli/compiledb.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+from qmk.commands import create_make_command
+from qmk.constants import QMK_FIRMWARE
+from qmk.decorators import automagic_keyboard, automagic_keymap
+
+
+@lru_cache(maxsize=10)
+def system_libs(binary: str):
+    """Find the system include directory that the given build tool uses.
+
+    Only tested on OSX+homebrew so far.
```
Comment (tzarc) [inline lib/python/qmk/cli/compiledb.py:25]:
```suggestion
```
Seems to be working fine on Linux:
```
% uname -a
Linux hailstorm 5.4.0-52-generic #57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
% qmk compiledb -kb massdrop/ctrl -km default
Ψ Making clean with make clean
Ψ Gathering build instructions from make -n massdrop/ctrl:default
Ψ Found 54 compile commands
Ψ Writing build database to /home/nickb/qmk_build/qmk_firmware/compile_commands.json
```

**Review by tzarc — CHANGES_REQUESTED** [review]
Just a minor docs change, otherwise LGTM.

**Comment by skullydazed** [comment]
Apologies for not getting back to this sooner. Overall this looks good but I wasn't able to come up with a name that worked better. Normally I'd have merged it anyway but 2020 has changed a lot of routines, and I hadn't cycled around to reviews again in a while.

In the meantime a lot of our current and future work has coalesced around a naming pattern for this sort of this. Could you make one last change to fit this scheme? Calling it `qmk generate-compile-commands` fits our existing scheme, and isn't too bad to tab complete.

**Comment by j-hui** [comment]
Was there anything else blocking this PR, aside from deciding on the name of the target? I'd also love to see this feature merged in!

**Pull request closed without merging**